File: Workspace/Personal Training/JSW/Baekjoon/4949/solution.py
import sys

stack = []
push = lambda x: stack.append(x)
pop = lambda: stack.pop() if stack else False
peek = lambda: stack[-1] if stack else False
pairs = {'(': ')', '[': ']'}
def isVPS(s):
    stack.clear()
    for c in s:
        if c in "([":
            push(c)
        elif c in ")]":
            popped = pop()
            if popped == False or pairs[popped] != c:
                return False
    return not stack

result = []
while (line:= sys.stdin.readline().strip('\n')) != '.':
    result.append('yes' if isVPS(line) else 'no')

print('\n'.join(result))

File: Workspace/Personal Training/JSW/Baekjoon/4949/solution.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;

Dictionary<char, char> wordPairs = new Dictionary<char, char>() {
    {'(', ')'},
    {'[', ']'},
    {'{', '}'}
};
Stack<char> stack = new Stack<char>();
bool isVPS(string str)  
{
    stack.Clear();
    foreach (char c in str)
    {
        if (wordPairs.ContainsKey(c)) stack.Push(c);
        if (wordPairs.ContainsValue(c))
        {
            if (stack.Count == 0) return false;
            var last = stack.Pop();
            if (wordPairs[last] != c) return false;
        }
    }

    return stack.Count == 0 ? true : false;
}

StringBuilder sb = new StringBuilder();
using (StreamReader sr = new StreamReader(Console.OpenStandardInput()))
{
    string line;
    while ((line = sr.ReadLine()) != ".")
    {
        sb.AppendLine(isVPS(line) ? "yes" : "no");
    }
}
Console.Write(sb);